contracts/stellar-interchain-token-service/src/migrate.rs

```diff
@@ -0,0 +1,101 @@
+#[derive(Clone, Debug, PartialEq, Eq)]
+pub struct CustomMigrationData {
+    pub upgrader_client: Address,
+    pub new_version: String,
+    pub token_ids: Vec<BytesN<32>>,
+    pub new_token_manager_wasm_hash: BytesN<32>,
+    pub new_interchain_token_wasm_hash: BytesN<32>,
+}
+
+pub mod legacy_storage {
+    use stellar_axelar_std::{contractstorage, contracttype, soroban_sdk, BytesN};
+
+    #[contractstorage]
+    enum LegacyDataKey {
+        #[temporary]
+        #[value(i128)]
+        FlowOut { flow_key: FlowKey },
+
+        #[temporary]
+        #[value(i128)]
+        FlowIn { flow_key: FlowKey },
+    }
+
+    #[contracttype]
+    #[derive(Clone, Debug)]
+    pub struct FlowKey {
+        pub token_id: BytesN<32>,
+        pub epoch: u64,
+    }
+}
+
+impl CustomMigratableInterface for InterchainTokenService {
+    type MigrationData = CustomMigrationData;
+    type Error = ContractError;
+
+    fn __migrate(env: &Env, migration_data: Self::MigrationData) -> Result<(), Self::Error> {
+        let CustomMigrationData {
+            upgrader_client,
+            new_version,
+            token_ids,
+            new_token_manager_wasm_hash,
+            new_interchain_token_wasm_hash,
+        } = migration_data;
+
+        storage::set_token_manager_wasm_hash(env, &new_token_manager_wasm_hash);
+        storage::set_interchain_token_wasm_hash(env, &new_interchain_token_wasm_hash);
+
+        let current_epoch = current_epoch(env);
+        let upgrader_client = UpgraderClient::new(env, &upgrader_client);
+
+        for token_id in token_ids.into_iter() {
+            let TokenIdConfigValue {
+                token_address: interchain_token,
+                token_manager,
+                token_manager_type,
+            } = storage::try_token_id_config(env, token_id.clone())
+                .ok_or(ContractError::InvalidTokenId)?;
+
+            upgrader_client.upgrade(
+                &token_manager,
+                &new_version,
+                &new_token_manager_wasm_hash,
+                &stellar_axelar_std::vec![env, ().into()],
+            );
+
+            if token_manager_type != TokenManagerType::LockUnlock {
+                upgrader_client.upgrade(
+                    &interchain_token,
+                    &new_version,
+                    &new_interchain_token_wasm_hash,
+                    &stellar_axelar_std::vec![env, ().into()],
+                );
+            }
+
+            let flow_key = legacy_storage::FlowKey {
+                token_id: token_id.clone(),
+                epoch: current_epoch,
+            };
+
+            if let Some(flow_out) = legacy_storage::try_flow_out(env, flow_key.clone()) {
+                storage::set_flow_out(env, token_id.clone(), current_epoch, &flow_out);
+            }
+            if let Some(flow_in) = legacy_storage::try_flow_in(env, flow_key.clone()) {
+                storage::set_flow_in(env, token_id.clone(), current_epoch, &flow_in);
+            }
+        }
```

Comment:
Stellar has pretty stringent read/write limit per txn
https://developers.stellar.org/docs/networks/resource-limits-fees

I'm concerned that we'll hit it with just a few token ids. A better approach is to simplify the atomic migration to just update the wasm hash so new tokens use the correct version, and defining a separate entrypoint `migrate_token_id` that migrates an existing token_id. There is no breaking interface between v1.0 and v1.1 I believe. This way, `migrate_token_id` can be called for each `token_id` without hitting any resource limits (gas, read/write, txn size etc.)